projecttacoma/measure-repository-service
# MRS as monorepo with frontend application

# Summary
This PR creates a frontend application for our existing Measure Repository Service by turning our existing measure repository service repo into a monorepo: the backend is one directory and the frontend is a separate directory. The frontend directory was created with [`create-next-app` ](https://github.com/vercel/next.js/tree/canary/packages/create-next-app) and the backend directory just contains what was previously in this repository, with a few exceptions that are now in the root directory. 

## New behavior
Now we have a Mrs. Frontend! Yay! If you cd into the frontend directory, you can then run commands as if this was its own repository. If you go back to the root directory, you can now run the server and the client at the same time using scripts in the root directory `package.json`. This will be explained more in the testing guidance, but in order to do this we used [npm workspaces](https://docs.npmjs.com/cli/v7/using-npm/workspaces).

NOTE: If you cd into the backend directory, all commands and resulting behavior should be the same as it was before.

## Code changes
Apologies for the massive code diff. Here are the important things to look at:
- `backend` - this contains the same code that we had before in this repository. Nothing in here has changed, so no need to worry about this.
- `frontend` - this was added using `create-next-app`. I would skim through the files but there's not a lot of functionality here yet.
- `.github` - I just kept this at the root and also made some changes to the `ci.yml` file. 
- `.gitignore` / `LICENSE` - Also kept these at the root.
- `package.json` / `package-lock.json` - New files for the root directory. Both the frontend and backend directories have their own. New scripts for running both the frontend and the backend at the same time made possible by npm workspaces.
- `README.md` - Short README for the overall repository.
- `tsconfig-base.json` - I added a tsconfig file to the root directory that contains the parts of the frontend and backend `tsconfig.json` that were the same. The `tsconfig.json` files in the frontend and backend directories now extend from the root file, `tsconfig-base.json`.

# Testing guidance
Test that the measure repository service is still working on its own:
1. `cd backend`
2. `npm run check`
3. `npm start`
4. Try some requests in Insomnia to make sure it is working.

Test the new frontend application:
1. `cd frontend`
2. `npm install`
3. `npm run dev`
4. Navigate to [http://localhost:3000](http://localhost:3000) and make sure you see the measure repository service frontend application.

Test the overall repository:
At the root directory:
1. `npm install`
2. `npm run start:all` - this starts both the frontend and backend at the same time. 
3. Navigate to [http://localhost:3000](http://localhost:3000) and make sure you see the measure repository service frontend application.
4. Try the other new scripts: `npm run check:all`, `npm run lint:all`, `npm run lint:all-fix`, `npm run prettier:all`, `npm run prettier:all-fix`
5. Let me know if you have any other ideas for any more scripts.

Other:
- Let me know if there is anything I should add or remove from any of the READMEs.
- Look at the scripts in the root `package.json`. Am I missing any that may be helpful? Do the names make sense?
